File: api/src/models/contracts/solutions.py
"""Pydantic types for Solutions — installable surfaces (success-criteria §3)."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, computed_field

SolutionScope = Literal["org", "global"]


class SolutionBase(BaseModel):
    slug: str = Field(min_length=1, max_length=255, description="Definition identity (shared across installs)")
    name: str = Field(min_length=1, max_length=255)
    global_repo_access: bool = False
    git_connected: bool = False
    git_repo_url: str | None = None
    repo_subpath: str | None = None
    git_ref: str | None = None


class SolutionCreate(SolutionBase):
    """Create-shape for an install.

    Install kind is DERIVED from ``organization_id``, per the unified --org
    standard — there is no ``scope`` input:

    - ``organization_id`` absent (HOME) => the caller's own org.
    - ``organization_id`` explicit null  => global (org NULL).
    - ``organization_id`` a UUID         => that org (cross-org admins).

    ``model_fields_set`` distinguishes absent (HOME) from explicit null (global).
    """

    organization_id: UUID | None = None


class SolutionUpdate(BaseModel):
    """Partial-update (PATCH) of an install's INSTALL-LOCAL fields only.

    ``slug`` is identity and is NOT editable here. Portable content
    (workflows/apps/forms/agents/tables/config declarations) is owned by the
    bundle/git and is read-only on this surface.

    PATCH semantics: ``organization_id=None`` is a legitimate value (global
    scope), so it is distinguished from "not provided" via
    ``model_fields_set`` — the endpoint applies only fields present in the
    request (``model_dump(exclude_unset=True)``).
    """

    name: str | None = None
    organization_id: UUID | None = None
    global_repo_access: bool | None = None
    git_connected: bool | None = None
    git_repo_url: str | None = None
    repo_subpath: str | None = None
    git_ref: str | None = None


class SolutionReadmeUpdate(BaseModel):
    """PUT body for an install's README markdown (Task 6).

    ``readme=None`` clears the README; a string sets it. The README is normally
    repo-sourced (deploy reads it from the repo-root README.md), but the UI can
    edit it directly on a disconnected install via this endpoint.
    """

    readme: str | None = None


class SolutionReadme(BaseModel):
    """GET/PUT response shape for an install's README markdown."""

    readme: str | None = None


class Solution(BaseModel):
    """Read-shape returned by REST.

    ``scope`` is DERIVED from ``organization_id`` (NULL == global), not stored
    on the ORM row — so it always reflects the install's true scope.
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    slug: str
    name: str
    organization_id: UUID | None = None
    global_repo_access: bool = False
    git_connected: bool = False
    git_repo_url: str | None = None
    # Subfolder within the connected repo holding this install's descriptor, and
    # the git ref (branch/tag) it tracks. Both None => repo root / default branch.
    repo_subpath: str | None = None
    git_ref: str | None = None
    # Version bookkeeping (Task 20): the deployed bundle's declared version and
    # what the last version-changing deploy replaced. Deploy-recorded, not
    # caller-settable — version rides in the BUNDLE (descriptor), not a request.
    version: str | None = None
    upgraded_from_version: str | None = None
    # Newest descriptor version available at the connected repo's ref HEAD when
    # it is PEP-440-greater than `version` (set by the update-check scheduler).
    # None => up to date / not git-connected / not yet checked. Drives the
    # "Update Available" badge.
    update_available_version: str | None = None
    # Persisted setup-completeness (Task 5/7): True when every required config
    # declaration has a matching Config value in the install's org scope.
    # Recomputed by install_zip after each deploy so it reflects the install's
    # state without a separate /setup call. Defaults True (no declarations = complete).
    setup_complete: bool = True

    @computed_field  # type: ignore[prop-decorator]
    @property
    def scope(self) -> SolutionScope:
        return "org" if self.organization_id is not None else "global"


class SolutionsList(BaseModel):
    solutions: list[Solution] = Field(default_factory=list)


class SolutionConfigStatus(BaseModel):
    """A config DECLARATION on an install, paired with whether a value is set in
    the install's org scope (values are instance-owned Config rows, never part of
    the declaration)."""

    id: UUID
    key: str
    type: str
    required: bool
    description: str | None = None
    value_set: bool


class SolutionEntitySummary(BaseModel):
    """Lightweight entity row for Solution-owned/capturable entity lists."""

    id: UUID
    name: str
    description: str | None = None
    organization_id: UUID | None = None
    slug: str | None = None
    path: str | None = None
    function_name: str | None = None
    type: str | None = None
    category: str | None = None
    access_level: str | None = None
    app_model: str | None = None
    is_active: bool | None = None
    logo: str | None = None
    source_table: str | None = None
    select: str | None = None
    created_at: datetime | None = None


class SolutionEntities(BaseModel):
    """Everything one install owns + its config declaration/value status."""

    solution: Solution
    workflows: list[SolutionEntitySummary] = Field(default_factory=list)
    apps: list[SolutionEntitySummary] = Field(default_factory=list)
    forms: list[SolutionEntitySummary] = Field(default_factory=list)
    agents: list[SolutionEntitySummary] = Field(default_factory=list)
    claims: list[SolutionEntitySummary] = Field(default_factory=list)
    tables: list[SolutionEntitySummary] = Field(default_factory=list)
    configs: list[SolutionConfigStatus] = Field(default_factory=list)
    required_configs_unset: list[str] = Field(default_factory=list)


class SolutionCaptureCandidates(BaseModel):
    """Loose same-scope entities that can be adopted into an install."""

    workflows: list[SolutionEntitySummary] = Field(default_factory=list)
    apps: list[SolutionEntitySummary] = Field(default_factory=list)
    forms: list[SolutionEntitySummary] = Field(default_factory=list)
    agents: list[SolutionEntitySummary] = Field(default_factory=list)
    claims: list[SolutionEntitySummary] = Field(default_factory=list)
    tables: list[SolutionEntitySummary] = Field(default_factory=list)
    configs: list[SolutionConfigStatus] = Field(default_factory=list)


# ── Dependency preview (capture + export) — §3.2/§3.3 ───────────────────────

# Entity kinds the dependency walker reasons about. ``module`` is a Python file
# under ``modules/`` (no DB row); the rest are DB entities keyed by id, except
# ``config`` which is keyed by its string key.
DependencyKind = Literal[
    "workflow", "table", "config", "form", "app", "agent", "module", "integration"
]


class DependencyRef(BaseModel):
    """One entity the walker pulled in or warned about.

    ``ref`` is the natural handle: a UUID for DB entities, a key for configs, a
    relative path for modules. ``name`` is the display label; ``in_selection``
    is true when the seed selection already includes this entity (so the UI can
    show it as "already selected" vs "will be pulled in").
    """

    kind: DependencyKind
    ref: str
    name: str
    in_selection: bool = False


class UnmetNeed(BaseModel):
    """One thing a solution needs that isn't satisfied in the install target.

    ``kind`` is "module" (a modules/*.py import not present in the bundle) or
    "solution_dep" (a cross-solution reference not installed). Surfaced at
    install/upgrade so the install can BLOCK rather than fail at runtime.
    """

    kind: Literal["module", "solution_dep"]
    ref: str
    detail: str | None = None


class OutsideReference(BaseModel):
    """An entity OUTSIDE the selection that references something INSIDE it.

    The capture/export preview surfaces these as non-blocking warnings: the
    referenced entity is being adopted by the install while ``referencer`` is
    left loose and will keep pointing at it across the scope boundary.
    """

    referencer_kind: DependencyKind
    referencer_ref: str
    referencer_name: str
    target_kind: DependencyKind
    target_ref: str
    target_name: str


class SolutionDependencyPreview(BaseModel):
    """What a capture/export selection actually grabs, for human review.

    ``pulled_in`` is the forward dependency closure beyond the seed selection
    (e.g. a captured workflow's ``modules/`` imports when ``include_imports`` is
    on, the tables/configs it reads, the workflow a captured form launches).
    ``outside_references`` are reverse-dependency warnings. The preview is the
    guard: every item is deselectable, nothing is silently blocked.
    """

    pulled_in: list[DependencyRef] = Field(default_factory=list)
    outside_references: list[OutsideReference] = Field(default_factory=list)
    # True when the static scan can't see everything (dynamic imports / computed
    # refs) — the UI nudges the human to add any missed file manually.
    scan_is_static: bool = True


class SolutionDependencyPreviewRequest(BaseModel):
    """Seed selection to preview, mirroring SolutionCaptureRequest's selectors."""

    workflows: list[UUID] = Field(default_factory=list)
    tables: list[UUID] = Field(default_factory=list)
    apps: list[UUID] = Field(default_factory=list)
    forms: list[UUID] = Field(default_factory=list)
    agents: list[UUID] = Field(default_factory=list)
    claims: list[UUID] = Field(default_factory=list)
    configs: list[str] = Field(default_factory=list)
    include_imports: bool = False


class SolutionRepoPreviewRequest(BaseModel):
    """Resolve an install plan from a git repo (+ optional subfolder/ref).
    Parse-only — no DB write, no S3, no build."""

    repo_url: str = Field(min_length=1, max_length=1024)
    repo_subpath: str | None = None
    git_ref: str | None = None
    # Install kind for install-from-repo, per the unified --org standard:
    # absent => the caller's own org (HOME); explicit null => global; a UUID =>
    # that org. The descriptor no longer carries scope.
    organization_id: UUID | None = Field(
        default=None,
        description="Target org for the install (absent => caller's org, null => global).",
    )


class SolutionEntityDiff(BaseModel):
    """Added/removed display names for ONE entity type in an upgrade preview.

    Identity is the deployer's per-install uuid5 remap (``solution_entity_id``),
    so "kept" (in neither list) means the deploy would UPDATE that row in place.
    """

    added: list[str] = Field(default_factory=list)
    removed: list[str] = Field(default_factory=list)


class SolutionConfigSchemaState(BaseModel):
    """The compared portion of a config declaration (type + required)."""

    type: str
    required: bool


class SolutionConfigSchemaChange(BaseModel):
    """One config declaration whose type/required changed between versions."""

    model_config = ConfigDict(populate_by_name=True)

    key: str
    from_: SolutionConfigSchemaState = Field(alias="from")
    to: SolutionConfigSchemaState


class SolutionConfigSchemaDiff(BaseModel):
    """Config DECLARATION diff (by key) for an upgrade preview."""

    added: list[str] = Field(default_factory=list)
    removed: list[str] = Field(default_factory=list)
    changed: list[SolutionConfigSchemaChange] = Field(default_factory=list)


class SolutionUpgradeDiff(BaseModel):
    """What deploying the previewed zip would change on the existing install."""

    workflows: SolutionEntityDiff = Field(default_factory=SolutionEntityDiff)
    tables: SolutionEntityDiff = Field(default_factory=SolutionEntityDiff)
    forms: SolutionEntityDiff = Field(default_factory=SolutionEntityDiff)
    agents: SolutionEntityDiff = Field(default_factory=SolutionEntityDiff)
    apps: SolutionEntityDiff = Field(default_factory=SolutionEntityDiff)
    claims: SolutionEntityDiff = Field(default_factory=SolutionEntityDiff)
    config_schemas: SolutionConfigSchemaDiff = Field(default_factory=SolutionConfigSchemaDiff)


class SolutionExistingInstall(BaseModel):
    """The install a previewed zip would UPGRADE (matched by slug + scope)."""

    id: UUID
    name: str
    version: str | None = None


class SolutionInstallPreview(BaseModel):
    """Parse-only preview of a Solution install zip — what it would create + its
    declared configs. Nothing is persisted by the preview endpoint.

    When an install already exists for the zip's slug at the requested scope,
    ``existing_install`` + ``diff`` describe the upgrade the install would
    perform (Task 22) — drag-drop routes to UPGRADE, never a second install.

    ``requires_password`` is True when the zip contains ``.bifrost/secrets.enc``
    (a full-backup export). The install endpoint requires a password to decrypt
    it; the UI should prompt for the password before the install POST."""

    slug: str | None = None
    name: str | None = None
    scope: SolutionScope | None = None
    version: str | None = None
    workflows: list[dict[str, Any]] = Field(default_factory=list)
    tables: list[dict[str, Any]] = Field(default_factory=list)
    apps: list[dict[str, Any]] = Field(default_factory=list)
    forms: list[dict[str, Any]] = Field(default_factory=list)
    agents: list[dict[str, Any]] = Field(default_factory=list)
    claims: list[dict[str, Any]] = Field(default_factory=list)
    config_schemas: list[dict[str, Any]] = Field(default_factory=list)
    # Each: {integration_name, template, position}. Secret-scrubbed skeletons
    # (no client_id/secret). Declared from integrations.get("X") refs.
    connection_schemas: list[dict[str, Any]] = Field(default_factory=list)
    existing_install: SolutionExistingInstall | None = None
    diff: SolutionUpgradeDiff | None = None
    requires_password: bool = False
    # Long-form README markdown read from the zip's repo-root README.md (Task 6).
    readme: str | None = None


class SolutionDeployRequest(BaseModel):
    """Full-replace deploy bundle for one install.

    ``python_files`` maps relative paths (e.g. ``workflows/w.py``,
    ``modules/x.py``) to UTF-8 source text, installed verbatim under the
    install's ``_solutions/{id}/`` prefix. ``workflows`` are manifest-shaped
    entity dicts to upsert (apps/forms/agents/tables join in later sub-plans).
    Deploy is non-interactive by contract — it always applies the full bundle.
    """

    python_files: dict[str, str] = Field(default_factory=dict)
    workflows: list[dict[str, Any]] = Field(default_factory=list)
    tables: list[dict[str, Any]] = Field(default_factory=list)
    # Each app: {id, slug, name, app_model, dependencies, access_level,
    # src_files: {rel: text} | dist_files: {rel: text},
    # bin_dist_files: {rel: base64}}. dist_files/bin_dist_files are the
    # disconnected fast-path (skip the server build); bin_dist_files carries
    # non-UTF-8 dist assets (images/fonts/wasm) base64-encoded so they survive
    # the round-trip unmangled.
    apps: list[dict[str, Any]] = Field(default_factory=list)
    # Each form: {id, name, description?, workflow_id?, fields: [...]}.
    forms: list[dict[str, Any]] = Field(default_factory=list)
    # Each agent: {id, name, system_prompt, description?, channels?, llm_model?}.
    agents: list[dict[str, Any]] = Field(default_factory=list)
    claims: list[dict[str, Any]] = Field(default_factory=list)
    # Each config schema: {id, key, type, required, description?, default?, position}.
    # DECLARATIONS only — never a value (values are instance-owned Config rows).
    config_schemas: list[dict[str, Any]] = Field(default_factory=list)
    # Each: {integration_name, template, position}. Secret-scrubbed skeletons
    # (no client_id/secret). Declared from integrations.get("X") refs.
    connection_schemas: list[dict[str, Any]] = Field(default_factory=list)
    # Each event/schedule trigger: a ManifestEventSource-shaped dict (source +
    # schedule/webhook config + nested subscriptions). Webhook instance secrets
    # are scrubbed; the instance re-establishes external state after install.
    events: list[dict[str, Any]] = Field(default_factory=list)
    # The bundle's declared version (bifrost.solution.yaml ``version:``).
    # Recorded on the install; an older version than installed is refused
    # unless ``force`` is set (Task 20 downgrade gate).
    version: str | None = None
    # Solution-level icon declared by ``logo:`` in bifrost.solution.yaml,
    # carried base64 by the CLI; deploy validates and stamps it on the install
    # (absent => cleared).
    logo_b64: str | None = None
    logo_content_type: str | None = None
    # Long-form README markdown sourced from the repo-root README.md (Task 6).
    # Deploy-owned full-replace: present => set, absent => cleared.
    readme: str | None = None
    force: bool = False


class SolutionDeleteSummary(BaseModel):
    """Counts of what a DELETE did. Pure-code entities (workflows/apps/forms/
    agents) and the install's config DECLARATIONS are deleted via DB cascade.
    Data-bearing entities are ORPHANED, not deleted: owned tables (and their
    documents) are detached and survive as ordinary org tables, and the
    install's config VALUES are stamped with orphan provenance and survive.
    The UI echoes these back to the operator."""

    solution_id: UUID
    workflows_deleted: int = 0
    apps_deleted: int = 0
    forms_deleted: int = 0
    agents_deleted: int = 0
    claims_deleted: int = 0
    config_declarations_deleted: int = 0
    tables_orphaned: int = 0
    config_values_orphaned: int = 0


class SolutionDeployResponse(BaseModel):
    solution_id: UUID
    workflows_upserted: int = 0
    workflows_deleted: int = 0
    tables_upserted: int = 0
    tables_deleted: int = 0
    apps_upserted: int = 0
    apps_deleted: int = 0
    forms_upserted: int = 0
    forms_deleted: int = 0
    agents_upserted: int = 0
    agents_deleted: int = 0
    claims_upserted: int = 0
    claims_deleted: int = 0
    # Empty integration shells created for declared connections that did not yet
    # exist globally (never clobbers a configured integration).
    integrations_shell_created: int = 0
    # Names of roles auto-created because the bundle referenced a role missing
    # from the target env. Created global + empty (grant nothing until assigned).
    # Surfaced so the operator sees them — and a typo'd role name is visible.
    roles_created: list[str] = Field(default_factory=list)


class SolutionCaptureRequest(BaseModel):
    """Move existing loose entities into an install in place.

    Entity ids must currently be unowned (``solution_id`` is null) and scoped
    the same way as the install. Config keys become declarations; their values
    stay in the install scope.
    """

    workflows: list[UUID] = Field(default_factory=list)
    tables: list[UUID] = Field(default_factory=list)
    apps: list[UUID] = Field(default_factory=list)
    forms: list[UUID] = Field(default_factory=list)
    agents: list[UUID] = Field(default_factory=list)
    claims: list[UUID] = Field(default_factory=list)
    configs: list[str] = Field(default_factory=list)
    include_imports: bool = Field(
        default=False,
        description=(
            "When false (default), bundle only the captured workflows' own "
            "source files. When true, also bundle the transitive import "
            "closure of `modules/` they reference (never the whole modules/ "
            "tree — only what is actually imported)."
        ),
    )


class SolutionCaptureResponse(BaseModel):
    solution_id: UUID
    workflows_captured: int = 0
    tables_captured: int = 0
    apps_captured: int = 0
    forms_captured: int = 0
    agents_captured: int = 0
    claims_captured: int = 0
    config_declarations_captured: int = 0


class PullAckEntity(BaseModel):
    """One captured entity the client has materialized into source `.bifrost/`."""

    entity_type: str  # table|form|agent|config|event|claim
    entity_id: str  # entity id; for config, its key


class PullAckRequest(BaseModel):
    """Tell the server which pending_captures rows `bifrost solution pull`
    materialized into source, so it can clear exactly those rows
    (server-authoritative — a stale client can't clear what it didn't pull)."""

    entities: list[PullAckEntity] = Field(default_factory=list)


class PullAckResponse(BaseModel):
    cleared: int = 0


class SolutionSetupItem(BaseModel):
    """One declared requirement paired with whether it's satisfied.

    For ``kind="config"``: a config declaration; ``is_set`` = a Config value
    exists in the install scope. For ``kind="connection"``: an integration
    declaration; ``is_set`` = a global Integration with that name exists.
    """

    key: str  # config key OR integration name
    type: str  # config value type for kind="config"; the literal "integration" for kind="connection"
    required: bool
    is_set: bool
    description: str | None = None
    default: str | None = None
    kind: Literal["config", "connection"] = "config"
    # Connection-only meta (defaults for config items):
    has_oauth: bool = False  # template carried OAuth shape — WARN-ONLY, never gates
    connected: bool = False  # informational: a token/mapping resolves


class SolutionSetupStatus(BaseModel):
    """Required-config setup status for a Solution install.

    ``setup_complete`` is True when every required declaration has a matching
    Config value in the install's org scope.
    """

    setup_complete: bool
    items: list[SolutionSetupItem]

File: client/src/services/solutions.ts
import { apiClient, authFetch } from "@/lib/api-client";
import { getErrorMessage } from "@/lib/api-error";
import type { components } from "@/lib/v1";

export type Solution = components["schemas"]["Solution"];
export type SolutionsList = components["schemas"]["SolutionsList"];
export type SolutionEntities = components["schemas"]["SolutionEntities"];
export type SolutionEntitySummary =
	components["schemas"]["SolutionEntitySummary"];
export type SolutionConfigStatus =
	components["schemas"]["SolutionConfigStatus"];
export type SolutionInstallPreview =
	components["schemas"]["SolutionInstallPreview"];
export type SolutionExistingInstall =
	components["schemas"]["SolutionExistingInstall"];
export type SolutionUpgradeDiff =
	components["schemas"]["SolutionUpgradeDiff"];
export type SolutionDeleteSummary =
	components["schemas"]["SolutionDeleteSummary"];
export type SolutionUpdate = components["schemas"]["SolutionUpdate"];
export type SolutionSetupStatus = components["schemas"]["SolutionSetupStatus"];
export type SolutionSetupItem = components["schemas"]["SolutionSetupItem"];
export type SolutionCaptureCandidates =
	components["schemas"]["SolutionCaptureCandidates"];
export type SolutionCaptureRequest =
	components["schemas"]["SolutionCaptureRequest"];
export type SolutionDependencyPreview =
	components["schemas"]["SolutionDependencyPreview"];
export type SolutionDependencyPreviewRequest =
	components["schemas"]["SolutionDependencyPreviewRequest"];
export type DependencyRef = components["schemas"]["DependencyRef"];
export type OutsideReference = components["schemas"]["OutsideReference"];
export type SolutionCaptureResponse =
	components["schemas"]["SolutionCaptureResponse"];
export type SolutionReadme = components["schemas"]["SolutionReadme"];
export type SolutionRepoPreviewRequest =
	components["schemas"]["SolutionRepoPreviewRequest"];

interface RequestOptions {
	signal?: AbortSignal;
}

export async function listSolutions(
	options: RequestOptions = {},
): Promise<SolutionsList> {
	const { signal } = options;
	const { data, error } = await apiClient.GET("/api/solutions", { signal });
	if (error) throw new Error(getErrorMessage(error, "Failed to list solutions"));
	return data;
}

export async function getSolution(
	solutionId: string,
	options: RequestOptions = {},
): Promise<Solution> {
	const { signal } = options;
	const { data, error } = await apiClient.GET("/api/solutions/{solution_id}", {
		params: { path: { solution_id: solutionId } },
		signal,
	});
	if (error) throw new Error(getErrorMessage(error, "Failed to get solution"));
	return data;
}

export async function getSolutionSetup(
	solutionId: string,
	options: RequestOptions = {},
): Promise<SolutionSetupStatus> {
	const { signal } = options;
	const { data, error } = await apiClient.GET(
		"/api/solutions/{solution_id}/setup",
		{ params: { path: { solution_id: solutionId } }, signal },
	);
	if (error) {
		throw new Error(getErrorMessage(error, "Failed to get solution setup status"));
	}
	return data;
}

export async function getSolutionReadme(
	solutionId: string,
	options: RequestOptions = {},
): Promise<SolutionReadme> {
	const { signal } = options;
	const { data, error } = await apiClient.GET(
		"/api/solutions/{solution_id}/readme",
		{ params: { path: { solution_id: solutionId } }, signal },
	);
	if (error) {
		throw new Error(getErrorMessage(error, "Failed to get solution readme"));
	}
	return data;
}

export async function putSolutionReadme(
	solutionId: string,
	readme: string | null,
	options: RequestOptions = {},
): Promise<SolutionReadme> {
	const { signal } = options;
	const { data, error } = await apiClient.PUT(
		"/api/solutions/{solution_id}/readme",
		{ params: { path: { solution_id: solutionId } }, body: { readme }, signal },
	);
	if (error) {
		throw new Error(getErrorMessage(error, "Failed to update solution readme"));
	}
	return data;
}

export async function getSolutionEntities(
	solutionId: string,
	options: RequestOptions = {},
): Promise<SolutionEntities> {
	const { signal } = options;
	const { data, error } = await apiClient.GET(
		"/api/solutions/{solution_id}/entities",
		{ params: { path: { solution_id: solutionId } }, signal },
	);
	if (error) {
		throw new Error(getErrorMessage(error, "Failed to get solution entities"));
	}
	return data;
}

export async function updateSolution(
	solutionId: string,
	update: SolutionUpdate,
	options: RequestOptions = {},
): Promise<Solution> {
	const { signal } = options;
	const { data, error } = await apiClient.PATCH(
		"/api/solutions/{solution_id}",
		{ params: { path: { solution_id: solutionId } }, body: update, signal },
	);
	if (error) throw new Error(getErrorMessage(error, "Failed to update solution"));
	return data;
}

/**
 * Trigger a pull/sync of a git-connected install (the "Update now" action).
 * Pulls the latest commit at the install's configured ref and re-applies the
 * solution.
 */
export async function syncSolution(
	solutionId: string,
	options: RequestOptions = {},
): Promise<void> {
	const { signal } = options;
	const { error } = await apiClient.POST("/api/solutions/{solution_id}/sync", {
		params: { path: { solution_id: solutionId } },
		signal,
	});
	if (error) throw new Error(getErrorMessage(error, "Failed to sync solution"));
}

/**
 * Preview a Solution install sourced from a git repository (parse-only). The
 * server clones the repo at the given ref/subpath and returns the same
 * `SolutionInstallPreview` shape as the zip-based `previewInstall`.
 */
export async function previewSolutionFromRepo(
	body: SolutionRepoPreviewRequest,
	options: RequestOptions = {},
): Promise<SolutionInstallPreview> {
	const { signal } = options;
	const { data, error } = await apiClient.POST(
		"/api/solutions/install/preview-repo",
		{ body, signal },
	);
	if (error) {
		throw new Error(getErrorMessage(error, "Failed to preview repository"));
	}
	return data;
}

/**
 * Install a Solution sourced from a git repository. The server clones the repo
 * at the given ref/subpath and installs it, returning the created `Solution`.
 */
export async function installSolutionFromRepo(
	body: SolutionRepoPreviewRequest,
	options: RequestOptions = {},
): Promise<Solution> {
	const { signal } = options;
	const { data, error } = await apiClient.POST(
		"/api/solutions/install/from-repo",
		{ body, signal },
	);
	if (error) {
		throw new Error(getErrorMessage(error, "Failed to install from repository"));
	}
	return data;
}

export async function getSolutionCaptureCandidates(
	solutionId: string,
	options: RequestOptions = {},
): Promise<SolutionCaptureCandidates> {
	const { signal } = options;
	const { data, error } = await apiClient.GET(
		"/api/solutions/{solution_id}/capture/candidates",
		{ params: { path: { solution_id: solutionId } }, signal },
	);
	if (error) {
		throw new Error(
			getErrorMessage(error, "Failed to list capture candidates"),
		);
	}
	return data;
}

export async function previewSolutionCapture(
	solutionId: string,
	request: SolutionDependencyPreviewRequest,
	options: RequestOptions = {},
): Promise<SolutionDependencyPreview> {
	const { signal } = options;
	const { data, error } = await apiClient.POST(
		"/api/solutions/{solution_id}/capture/preview",
		{ params: { path: { solution_id: solutionId } }, body: request, signal },
	);
	if (error) {
		throw new Error(getErrorMessage(error, "Failed to preview capture"));
	}
	return data;
}

export async function captureSolutionEntities(
	solutionId: string,
	request: SolutionCaptureRequest,
	options: RequestOptions = {},
): Promise<SolutionCaptureResponse> {
	const { signal } = options;
	const { data, error } = await apiClient.POST(
		"/api/solutions/{solution_id}/capture",
		{ params: { path: { solution_id: solutionId } }, body: request, signal },
	);
	if (error) {
		throw new Error(getErrorMessage(error, "Failed to capture entities"));
	}
	return data;
}

/**
 * Set a config VALUE for a Solution install's org scope. Config values are
 * instance-owned `Config` rows (never part of the portable declaration), so we
 * write them through the existing `/api/config` endpoint scoped to the install's
 * organization. `organizationId` is the install's org (`null` for a global install).
 */
export async function setSolutionConfig(
	params: {
		key: string;
		value: string;
		type: components["schemas"]["ConfigType"];
		organizationId: string | null;
	},
	options: RequestOptions = {},
): Promise<void> {
	const { key, value, type, organizationId } = params;
	const { error } = await apiClient.POST("/api/config", {
		body: { key, value, type, organization_id: organizationId },
		signal: options.signal,
	});
	if (error) throw new Error(getErrorMessage(error, "Failed to save config value"));
}

export async function deleteSolution(
	solutionId: string,
	options: RequestOptions = {},
): Promise<SolutionDeleteSummary> {
	const { signal } = options;
	const { data, error } = await apiClient.DELETE(
		"/api/solutions/{solution_id}",
		{ params: { path: { solution_id: solutionId } }, signal },
	);
	if (error) throw new Error(getErrorMessage(error, "Failed to delete solution"));
	return data;
}

/**
 * Download the install's workspace zip. Defaults to "shareable" mode (strips
 * secrets). Pass mode="full" with a password to produce an encrypted full
 * backup that includes secret config values.
 */
export async function exportSolution(
	solutionId: string,
	mode: "shareable" | "full" = "shareable",
	password?: string,
	includeData?: boolean,
): Promise<{ blob: Blob; filename: string }> {
	// POST so the full-backup password rides in the request body, not the URL
	// query string (a query-string secret leaks into logs/proxies/history).
	// mode + include_data are not sensitive and stay in the query.
	const params = new URLSearchParams({ mode });
	if (includeData) params.set("include_data", "true");
	const response = await authFetch(
		`/api/solutions/${solutionId}/export?${params.toString()}`,
		{
			method: "POST",
			headers: { "Content-Type": "application/json" },
			body: JSON.stringify(password ? { password } : {}),
		},
	);
	if (!response.ok) {
		throw new Error(
			await parseUploadError(response, "Failed to export solution"),
		);
	}
	const disposition = response.headers.get("Content-Disposition") ?? "";
	const match = /filename="([^"]+)"/.exec(disposition);
	return {
		blob: await response.blob(),
		filename: match?.[1] ?? `solution-${solutionId}.zip`,
	};
}

async function parseUploadError(
	response: Response,
	fallback: string,
): Promise<string> {
	const body = await response.json().catch(() => ({}));
	if (body && typeof body.detail === "string") {
		return body.detail;
	}
	return fallback;
}

/**
 * Preview a Solution install zip (parse-only). Posts a multipart `file` and an
 * optional `organization_id` (empty/absent = global) so the server can match
 * an existing install at that scope and return `existing_install` + `diff`.
 */
export async function previewInstall(
	file: File,
	params: { organizationId?: string } = {},
	options: RequestOptions = {},
): Promise<SolutionInstallPreview> {
	const formData = new FormData();
	formData.append("file", file);
	formData.append("organization_id", params.organizationId ?? "");

	const response = await authFetch("/api/solutions/install/preview", {
		method: "POST",
		body: formData,
		signal: options.signal,
	});
	if (!response.ok) {
		throw new Error(
			await parseUploadError(
				response,
				`Failed to preview install: ${response.statusText}`,
			),
		);
	}
	return response.json();
}

/**
 * Install a Solution zip. Posts a multipart `file`, optional `organization_id`
 * (empty string installs globally), and `config_values` (JSON-encoded map).
 * Pass `force: true` to override the server's downgrade guard (409 when the
 * package version is older than the installed version).
 *
 * `replaceSecrets: true` — re-install overwriting existing secret config
 *   values (send when the user confirms the collision prompt on 409).
 * `replaceData: true` — re-install overwriting existing table data.
 *
 * NOTE: `password` for full-backup installs is NOT yet wired into the install
 * UI. The server will return 422 if a full-backup zip is uploaded without the
 * correct password field. A future task should add a password prompt to the
 * install flow (CreateEditSolution / update dialog) when the preview response
 * indicates a full-backup zip. For now, the gap is documented here and in the
 * SolutionDetail import-collision handler.
 */
export async function installSolution(
	params: {
		file: File;
		organizationId?: string;
		configValues?: Record<string, unknown>;
		force?: boolean;
		replaceSecrets?: boolean;
		replaceData?: boolean;
		password?: string;
	},
	options: RequestOptions = {},
): Promise<Solution> {
	const {
		file,
		organizationId,
		configValues,
		force,
		replaceSecrets,
		replaceData,
		password,
	} = params;
	const formData = new FormData();
	formData.append("file", file);
	formData.append("organization_id", organizationId ?? "");
	formData.append("config_values", JSON.stringify(configValues ?? {}));
	if (replaceSecrets) formData.append("replace_secrets", "true");
	if (replaceData) formData.append("replace_data", "true");
	if (password) formData.append("password", password);

	const url = force
		? "/api/solutions/install?force=true"
		: "/api/solutions/install";
	const response = await authFetch(url, {
		method: "POST",
		body: formData,
		signal: options.signal,
	});
	if (!response.ok) {
		const err = new Error(
			await parseUploadError(
				response,
				`Failed to install solution: ${response.statusText}`,
			),
		);
		// Attach status so callers can branch on 409 / 422 without re-parsing.
		(err as Error & { status?: number }).status = response.status;
		throw err;
	}
	return response.json();
}
